File: brain/serial-client.rb
# -*- coding: utf-8 -*-
require 'rubygems'
require 'serialport'
#require 'httpclient'
require 'digest/md5'
require "net/http"
require "uri"

if ARGV.size >= 1
  serial = ARGV[0]
  puts serial
  url = ARGV[1]
  sp = SerialPort.new(serial, 9600)
else
  sp = nil
  puts "serial_client.rb /dev/ttyACM1 http://sushi.yuiseki.net/brain"
  exit
end

def post(body, url)
  begin
    uri = URI.parse(url)
    Net::HTTP.start(uri.host, uri.port){|http|
      resp = http.post(uri.path, body)
    }
  rescue Errno:: ECONNREFUSED
    puts "error"
  end
end

while true do
  line = sp.gets
  puts line
  csv = Time.now.to_i.to_s+','+line
  post(csv, url)
end

File: brain/serial-client.pl
#!/usr/bin/perl

use strict;
use LWP::UserAgent;
use UUID::Random;
use constant file_username => ".serial-client.username";

my $DEBUG=0;

my $comport = "/dev/ttyS6";
my $posturl = 'http://sushi.yuiseki.net:4444/brain';
my $userid  = getusername();


$posturl = $ARGV[0] || $posturl;

printf("\033[1;32mhttp://sushi.yuiseki.net:4444/brain?userid=%s&\033[0m\n\n", &urlencode($userid));







my $ua = new LWP::UserAgent;
open(FH, "<", $comport) or die("cannot open ", $comport);

while(<FH>)
{
  if(/ERROR/){ next; }
  my $line = $_;
  my @elements = split(/,/, $line);
  chop $line; 

  if(100 < $elements[0]){ 
    print "Low quality signal: ", $line , "\n";
    next;
  }

  my $postdata = sprintf("%ld,%s,%s", time(), $userid, $line);

  my $req = new HTTP::Request(POST => $posturl);
  $req->content_type("text/csv");
  $req->content($postdata);

  print $postdata,"\n";

  my $res = $ua->request($req);
  
  if($res->is_success){
    print $res->code, " - ", $res->content if $DEBUG
  }
  else{
    print STDERR "error: ", $res->code, "\n";
  }
}






sub
getusername
{
  my $file_username = file_username;
  my $username;

  if(-f $file_username)
  {
    open(FH, $file_username) or die($file_username);
    my @content = <FH>;
    close(FH);
    return join('', @content);
  }
  else{
    my $uuid = UUID::Random::generate;
    print "my new userid: $uuid\n" ;
    open(FH, ">" . $file_username) or die($file_username);
    print FH $uuid;
    close(FH);
    return $uuid;
  }
}


sub
urlencode($)
{
  my ($str) = @_;
  $str =~ s/([^\w ])/'%'.unpack('H2', $1)/eg;
  $str =~ tr/ /+/;
  return $str;
}


